### ISSUE TITLE
Using one more cpu archs at the same time

### ISSUE THREAD

**Issue by lieanu** [bdcht/amoco#21]
Hi, 

I got confused by the following codes, maybe this is an Amoco bug.

The code as follows:

https://gist.github.com/lieanu/f65788ff947a04d50aa0

``` python

import amoco
import amoco.system.raw
import amoco.system.core

def sym_exec_gadget_and_get_mapper(code, cpu):
    '''Taken from https://github.com/0vercl0k/stuffz/blob/master/look_for_gadgets_with_equations.py'''

    p = amoco.system.raw.RawExec(
        amoco.system.core.DataIO(code), cpu
    )

    try:
        blocks = list(amoco.lsweep(p).iterblocks())
    except:
        return None

    if len(blocks) == 0:
        return None

    mp = amoco.cas.mapper.mapper()
    for block in blocks:
        if block.instr[-1].mnemonic.lower() == 'call':
            p.cpu.i_RET(None, block.map)

        try:
            mp >>= block.map
        except Exception as e:
            pass

    return mp

if __name__ == "__main__":

    # pop rdi; ret --> "\x5f\xc3"
    print "-"*20, "AMD64", "-"*20
    print "Instr: ", "pop rdi; ret"
    import amoco.arch.x64.cpu_x64 as amd64_cpu
    cpu = amd64_cpu
    print sym_exec_gadget_and_get_mapper("\x5f\xc3", cpu)

    # pop eax; ret --> "\x58\xc3"
    print "-"*20, "I386", "-"*20
    print "Instr: ", "pop eax; ret"
    import amoco.arch.x86.cpu_x86 as i386_cpu
    cpu = i386_cpu
    print sym_exec_gadget_and_get_mapper("\x58\xc3", cpu)

    # pop rdi; ret --> "\x5f\xc3"
    print "-"*20, "AMD64 again", "-"*20
    print "Instr: ", "pop rdi; ret"
    import amoco.arch.x64.cpu_x64 as amd64_cpu
    cpu = amd64_cpu
    print sym_exec_gadget_and_get_mapper("\x5f\xc3", cpu)
```

The outputs:

```
-------------------- AMD64 --------------------
Instr:  pop rdi; ret
rdi <- { | [0:64]->M64(rsp) | } instructions [352]                              
rsp <- { | [0:64]->(rsp+0x10) | }
rip <- { | [0:64]->M64(rsp+8) | }
-------------------- I386 --------------------
Instr:  pop eax; ret
eax <- { | [0:32]->M32(esp) | } instructions [352]                              
esp <- { | [0:32]->(esp+0x8) | }
eip <- { | [0:32]->M32(esp+4) | }
-------------------- AMD64 again --------------------
Instr:  pop rdi; ret
rdi <- { | [0:64]->M64(esp) | }
esp <- { | [0:32]->(esp+0xc) | }
eip <- { | [0:32]->M32(esp+8) | }
```

You can see that the mapper of `pop rdi; ret` after switching to amd64 cpu, is incorrect.

Regards,
lieanu

**Comment by zachriggle** [comment]
Yeah, it looks like it gets confused after the first pop.  It even changes register names and word sizes.

From your output:

```
rdi <- { | [0:64]->M64(esp) | }
esp <- { | [0:32]->(esp+0xc) | }
eip <- { | [0:32]->M32(esp+8) | }
```

ESP and EIP should be RSP and RIP.  +c should be +10 and +8 is somehow correct?

**Comment by bdcht** [comment]
Hi,
indeed it's a known limitation of amoco due to some "globals" set when a
cpu module is imported. As far as I guess,
in your example the last import does nothing because Python knows the
module was already imported before, but meanwhile some core variables have
been scratched by the import of cpu_x64.

It's a limitation by design for now
if you need to study several archs, just use several interpreters. still,
the use case for say 2 archs in the same session makes sense for
translating or comparing semantics, so it will be fixed in the future...

**Comment by lieanu** [comment]
Thanks for your reply, I will close this issue temporarily.

**Comment by Nordwald** [comment]
Any chances of getting this fixed?

The support for multiple architectures is a major feature of amoco, sadly this bug hampers the usability.

**Comment by bdcht** [comment]
its quite a big fix unfortunately, still it is indeed planned to be fixed in 2.4.7 with python3 support, probably out in december...
